src/psbt.c

```diff
@@ -0,0 +1,2302 @@
+                if (!clone_bytes((unsigned char **)&result->items[i].origin.path, (unsigned char *)keypaths->items[i].origin.path, keypaths->items[i].origin.path_len * sizeof(*keypaths->items[i].origin.path))) {
+                    ret = WALLY_ENOMEM;
+                }
+                result->items[i].origin.path_len = keypaths->items[i].origin.path_len;
+            }
+
+            if (ret != WALLY_OK) {
+                keypath_map_free(result);
+            }
+        }
+        result->num_items = keypaths->num_items;
+    }
+    return ret == WALLY_OK ? result : NULL;
+}
+
+int add_new_keypath(struct keypath_map *keypaths,
+    unsigned char pubkey[65],
+    unsigned char fingerprint[4],
+    uint32_t *path,
+    size_t path_len)
+{
+    size_t latest;
+
+    if (keypaths->num_items == keypaths->items_allocation_len) {
+        if (keypaths->items_allocation_len == 0) {
+            keypaths->items_allocation_len = 1;
+        }
+        keypaths->items_allocation_len *= 2;
+        struct keypath_map *new_keypaths = clone_keypath_map(keypaths);
+        if (!new_keypaths) {
+            return WALLY_ENOMEM;
+        }
+        keypaths->items = new_keypaths->items;
+        keypaths->num_items = new_keypaths->num_items;
+        keypaths->items_allocation_len = new_keypaths->items_allocation_len;
+        wally_free(new_keypaths);
+    }
+
+    latest = keypaths->num_items;
+
+    if (!memcpy(&keypaths->items[latest].pubkey, pubkey, 65) ||
+        !memcpy(&keypaths->items[latest].origin.fingerprint, fingerprint, 4)) {
+        return WALLY_ENOMEM;
+    }
+    if (path) {
+        if (!clone_bytes((unsigned char **)&keypaths->items[latest].origin.path, (unsigned char *)path, path_len * sizeof(*path))) {
+            return WALLY_ENOMEM;
+        }
+        keypaths->items[latest].origin.path_len = path_len;
+    }
+    keypaths->num_items++;
+
+    return WALLY_OK;
+}
+
+static int add_keypath_item(struct keypath_map *keypaths, struct keypath_item *item)
+{
+    return add_new_keypath(keypaths, item->pubkey,
+                            item->origin.fingerprint,
+                            item->origin.path,
+                            item->origin.path_len);
+}
+
+int partial_sigs_map_init_alloc(size_t alloc_len, struct partial_sigs_map **output)
+{
+    struct partial_sigs_map *result;
+
+    TX_CHECK_OUTPUT;
+    TX_OUTPUT_ALLOC(struct partial_sigs_map);
+
+    if (alloc_len) {
+        result->items = wally_malloc(alloc_len * sizeof(*result->items));
+        if (!result->items) {
+            wally_free(result);
+            *output = NULL;
+            return WALLY_ENOMEM;
+        }
+        wally_clear(result->items, alloc_len * sizeof(*result->items));
+    }
+    result->items_allocation_len = alloc_len;
+    result->num_items = 0;
+    return WALLY_OK;
+}
+
+int partial_sigs_map_free(struct partial_sigs_map *sigs)
+{
+    size_t i;
+
+    if (sigs) {
+        if (sigs->items) {
```

Comment:
the two if statements could be merged in a single one

Reply:
Done here and in the other `clone_*_map` functions.